bundles/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/radiora/RS232Connection.java

```diff
@@ -42,32 +39,42 @@
 
     private final Logger logger = LoggerFactory.getLogger(RS232Connection.class);
 
+    protected SerialPortManager serialPortManager;
     protected SerialPort serialPort;
 
     protected BufferedReader inputReader;
 
     protected RadioRAFeedbackListener listener;
     protected RS232MessageParser parser = new RS232MessageParser();
 
+    public RS232Connection(SerialPortManager serialPortManager) {
+        super();
+        this.serialPortManager = serialPortManager;
+    }
+
     @Override
     public void open(String portName, int baud) throws RadioRAConnectionException {
-        CommPortIdentifier commPort = null;
+        SerialPortIdentifier portIdentifier;
 
-        try {
-            commPort = CommPortIdentifier.getPortIdentifier(portName);
-        } catch (NoSuchPortException e) {
-            logAvailablePorts();
+        // Exit if no identifiers exist to work around possible library bug. May not be needed in 3.0.
```

Comment:
I would just get the identifier and check if it is null (the code below). Recently I looked into rfc2217 connections (https://github.com/openhab/openhab-core/issues/1462) and these will not be returned in this list.

Reply:
Are you sure this bug work-around is no longer needed? If so I can remove it from two places here and one in the new alarmdecoder binding.

Reply:
Do you know what bug is worked around? I haven't noticed any bugs with bindings that don't use it ([dsmr](https://github.com/openhab/openhab-addons/blob/48dc8fb3099f2db82faf1ee1805ff48ac34fc56a/bundles/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/connector/DSMRSerialConnector.java#L109-L126), [plugwise](https://github.com/openhab/openhab-addons/blob/48dc8fb3099f2db82faf1ee1805ff48ac34fc56a/bundles/org.openhab.binding.plugwise/src/main/java/org/openhab/binding/plugwise/internal/PlugwiseCommunicationContext.java#L117)). The way it's written now introduces a bug for sure because [RFC2217PortProvider.getSerialPortIdentifiers()](https://github.com/openhab/openhab-core/blob/178ffdf9c275fd086f76f02d6be64ca29edbdc35/bundles/org.openhab.core.io.transport.serial.rxtx.rfc2217/src/main/java/org/openhab/core/io/transport/serial/rxtx/rfc2217/internal/RFC2217PortProvider.java#L50-L54) will not return any ports so if someone wants to use an rfc2217 connection it will stop because it thinks it doesn't exist.

Reply:
A possible JVM crash, supposedly. I saw similar code in the zigbee  or zwave binding with a comment that said it was necessary to avoid a bug in the serial library.

Reply:
Another reason for not checking the `getSerialPortIdentifiers()` is that the transport will also add ports to the `gnu.io.rxtx.SerialPorts` environment variable. So that way you can use non-standard serial ports (e.g. from udev rules) that are not discovered by the nrjavaserial library (see [code](https://github.com/openhab/openhab-core/blob/178ffdf9c275fd086f76f02d6be64ca29edbdc35/bundles/org.openhab.core.io.transport.serial.rxtx/src/main/java/org/openhab/core/io/transport/serial/internal/SerialPortUtil.java#L51-L63)) without having to manually add them to `gnu.io.rxtx.SerialPorts`.

Reply:
The [API docs](https://github.com/openhab/openhab-core/blob/178ffdf9c275fd086f76f02d6be64ca29edbdc35/bundles/org.openhab.core.io.transport.serial/src/main/java/org/openhab/core/io/transport/serial/SerialPortProvider.java#L47-L54) also state that these methods only return discovered ports (which may be none at all depending on the underlying serial implementation). So I think it's best to only use them for showing available ports in UIs and logging etc.

I think it would be better if the zwave/zigbee bindings just try to open the port to see if a  [device is unplugged](https://github.com/openhab/org.openhab.binding.zwave/blob/bfc26bb7c5f1e0b3fc605edca6cb48366786528f/src/main/java/org/openhab/binding/zwave/handler/ZWaveSerialHandler.java#L100-L101) @cdjackson. If that throws an exception they aren't plugged in. This will also solve detection issues and make these bindings more compatible with rfc2217.

Reply:
I'm not quite sure why we are discussing the zwave/zigbee bindings here, but...

> I think it would be better if the zwave/zigbee bindings just try to open the port to see if a device is unplugged @cdjackson.

The referenced code is only from the ZWave binding. I'm not familiar with why this was added - it was added by @bodiroga as part of the serial port reconnection changes. I don't mind if someone wants to change this logic - it will be changed in future as the binding is being rewritten.

Reply:
Thanks @cdjackson. I think we're just interested in determining whether that work-around code (or preferable some other fix in place of it) is still necessary. I don't really know the history of it. I just saw the comment "in some rare cases we have to check whether a port really exists, because if it doesn't the call to CommPortIdentifier#open will kill the whole JVM" in the Zigbee binding when I was looking for examples of using SerialPortManager:

https://github.com/openhab/org.openhab.binding.zigbee/blob/fac0cb77e61edebf8a9b1204879314f710b1f5bb/org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/handler/ZigBeeSerialPort.java#L139-L145

I thought that a JVM crash sounded undesirable :-). The work-around seemed relatively harmless and inexpensive, so I included it in the alarmdecoder binding, and later in the lutron binding. I'm not sure if any others have done the same thing. Now that @wborn has discovered the workaround prevents RFC2217 connections from working, it seems advisable to get rid of it.  But only if the original JVM-crashing bug is gone. In the Zigbee code it looks like that commit containing it came from @triller-telekom, so maybe he knows the history of it?

Reply:
That may be linked in some way to the USB discovery - I'm not sure. Maybe Stefan can help, although I think they have now forked the binding and are not directly using this code.  We were going to remove the USB discovery as it [caused other issues with device names](https://github.com/openhab/org.openhab.binding.zigbee/issues/307).

Reply:
I think we can safely remove it from this binding anyways. It will result in identical behavior as the [original code using gnu.io](https://github.com/openhab/openhab-addons/blob/48dc8fb3099f2db82faf1ee1805ff48ac34fc56a/bundles/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/radiora/RS232Connection.java#L53-L88) also just gets the identifier and opens the port. I couldn't find any other bindings doing this workaround (except for alarmdecoder). So if it is a big issue, it would have been reported for many other bindings by now.

Reply:
Ok. I'll go ahead and remove it now. I'd still be interested to hear about what the original problem was, though.

Reply:
Yes I'd also like to know. If it still exists we can reproduce the issue and fix the root cause. :-)

Reply:
Also removed from the alarmdecoder binding in #7615.